File: python3/133-clone-graph.py
from typing import Optional
from collections import defaultdict

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

# # helper function to build adjacency list
# def build_adj_list(node: Optional['Node']):

#     D = defaultdict(list)

#     def dfs_visit(node: Optional['Node']):
#         if not node:
#             return
#         for neighbor in node.neighbors:
#             D[node.val].append(neighbor.val)
#             dfs_visit(neighbor)

#     dfs_visit(node)
#     return D

def cloneGraph(node: Optional['Node']) -> Optional['Node']:
    
    # empty init graph
    if not node:
        return None

    # maintain a hashmap to store visited nodes
    visited = defaultdict()
    
    def dfs_visit(node: Optional['Node']) -> Optional['Node']:
        
        # stop exploration if the node was already visited
        if node.val in visited:
            return visited[node.val]
        
        # create a new copy of the current node and store the object in visited hashmap
        copy = Node(node.val)
        visited[node.val] = copy

        # visit all neighbor nodes
        for nei in node.neighbors:
            nei_copy = dfs_visit(nei)
            copy.neighbors.append(nei_copy)

        # return copy of reference to cloned graph
        return copy
    
    # driver
    return dfs_visit(node)


graph = Node(1, [Node(2, [Node(3)]), Node(4, [Node(5)])])
new_graph = cloneGraph(graph)

File: cpp/133-clone-graph.cpp
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

// Definition for a Node.
class Node
{
public:
    int val;
    vector<Node *> neighbors;
    Node()
    {
        val = 0;
        neighbors = vector<Node *>();
    }
    Node(int _val)
    {
        val = _val;
        neighbors = vector<Node *>();
    }
    Node(int _val, vector<Node *> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution
{
public:
    Node *bfsHelper(Node *node)
    {
        // maintain a hashmap of node objects where key is original, value is cloned
        // bfs approach
        unordered_map<Node *, Node *> visited;
        queue<Node *> q;

        // init
        Node *clone = new Node(node->val);
        visited[node] = clone;
        q.push(node);

        while (!q.empty())
        {
            Node *curr_node = q.front();
            q.pop();

            for (Node *adj_node : curr_node->neighbors)
            {
                // if adjacent node not visited, clone and store it in the hashmap
                if (visited.find(adj_node) == visited.end())
                {
                    visited[adj_node] = new Node(adj_node->val);
                    q.push(adj_node);
                }
                // push it to the neighbors of the current node
                visited[curr_node]->neighbors.push_back(visited[adj_node]);
            }
        }
        return clone;
    }

    Node *dfsHelper(Node *node, unordered_map<Node *, Node *> &visited)
    {
        if (visited.find(node) != visited.end())
        {
            return visited[node];
        }
        Node *clone = new Node(node->val);
        visited[node] = clone;

        for (Node *adj_node : node->neighbors)
        {
            Node *adj_clone = dfsHelper(adj_node, visited);
            visited[node]->neighbors.push_back(adj_clone);
        }
        return clone;
    }
    Node *cloneGraph(Node *node)
    {
        if (!node)
        {
            return nullptr;
        }
        unordered_map<Node *, Node *> visited;
        return bfsHelper(node);
    }
};